ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcStubPool.java

```diff
@@ -0,0 +1,105 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.grpc.server;
+
+import org.apache.ratis.protocol.RaftPeer;
+import org.apache.ratis.thirdparty.io.grpc.ManagedChannel;
+import org.apache.ratis.thirdparty.io.grpc.netty.NettyChannelBuilder;
+import org.apache.ratis.thirdparty.io.grpc.stub.AbstractStub;
+import org.apache.ratis.thirdparty.io.netty.channel.ChannelOption;
+import org.apache.ratis.thirdparty.io.netty.channel.WriteBufferWaterMark;
+import org.apache.ratis.thirdparty.io.netty.channel.nio.NioEventLoopGroup;
+import org.apache.ratis.thirdparty.io.netty.channel.socket.nio.NioSocketChannel;
+
+import java.util.List;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+
+final class GrpcStubPool<S extends AbstractStub<S>> implements AutoCloseable {
+
+    static final class PooledStub<S extends AbstractStub<S>> {
+        private final ManagedChannel ch;
+        private final S stub;
+        private final Semaphore permits;
+        PooledStub(ManagedChannel ch, S stub, int maxInflight) {
+            this.ch = ch;
+            this.stub = stub;
+            this.permits = new Semaphore(maxInflight);
+        }
+
+        S getStub() {
+            return stub;
+        }
+    }
+
+    private final List<PooledStub<S>> pool;
+    private final AtomicInteger rr = new AtomicInteger();
+    private final NioEventLoopGroup elg;
+    private final int size;
+
+    GrpcStubPool(RaftPeer target, int n, java.util.function.Function<ManagedChannel, S> stubFactory) {
+        this(target, n, stubFactory, Math.max(2, Runtime.getRuntime().availableProcessors()/2), 16);
+    }
+
+    GrpcStubPool(RaftPeer target, int n,
+                 java.util.function.Function<ManagedChannel, S> stubFactory,
+                 int elgThreads, int maxInflightPerConn) {
+        this.elg = new NioEventLoopGroup(elgThreads);
+        java.util.ArrayList<PooledStub<S>> tmp = new java.util.ArrayList<>(n);
+        for (int i = 0; i < n; i++) {
+            ManagedChannel ch = NettyChannelBuilder.forTarget(target.getAddress())
+                    .eventLoopGroup(elg)
+                    .channelType(NioSocketChannel.class)
+                    .keepAliveTime(30, java.util.concurrent.TimeUnit.SECONDS)
+                    .keepAliveWithoutCalls(true)
+                    .idleTimeout(24, java.util.concurrent.TimeUnit.HOURS)
+                    .withOption(ChannelOption.WRITE_BUFFER_WATER_MARK, new WriteBufferWaterMark(64<<10, 128<<10))
+                    .usePlaintext()
+                    .build();
```

Comment:
It needs the TLS parameter.

BTW, how did you come up with the other parameters?

Reply:
Updated.

Mostly recommended by GPT while testing the performance.